strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,232 @@
+
+    # pylint: disable=arguments-differ, invalid-overridden-method
+    def context(self, *args, copies=1, shift="end"):
+        input_check(args, copies)
+        self.copies = copies
+        self.tdm_params = args
+        self.shift = shift
+        self.loop_vars = self.params(*[f"p{i}" for i in range(len(args))])
+        return self
+
+    def __enter__(self):
+        super().__enter__()
+        return self.loop_vars, self.register
+
+    def __exit__(self, ex_type, ex_value, ex_tb):
+        super().__exit__(ex_type, ex_value, ex_tb)
+        self.spatial_modes = validate_measurements(self.circuit, self.N)
+        self.construct_circuit()
+
+    def construct_circuit(self):
+        """Construct program with the register shift"""
+        cmds = self.circuit.copy()
+        self.circuit = []
+
+        self.timebins = len(self.tdm_params[0])
+        self.total_timebins = self.timebins * self.copies
+
+        q = self.register
+
+        sm = []
+        for i in range(len(self.N)):
+            start = sum(self.N[:i])
+            stop = sum(self.N[:i]) + self.N[i]
+            sm.append(slice(start, stop))
+        # Above we define a list of slice intervals;
+        # each slice interval corresponding to one spatial mode
+
+        # For instance, say
+        # N = [5, 4, 9, 1],
+        # then this corresponds to
+        # 5 concurrent modes in spatial mode A
+        # 4 concurrent modes in spatial mode B
+        # 9 concurrent modes in spatial mode C
+        # 1 concurrent modes in spatial mode D.
+
+        # The entries of sm will then be:
+        # slice(0, 6, None)   for spatial mode A
+        # slice(6, 10, None)  for spatial mode B
+        # slice(10, 19, None) for spatial mode C
+        # slice(19, 20, None) for spatial mode D.
+
+        # This helps to define:
+        # q[sm[0]] as concurrent modes of spatial mode A
+        # q[sm[1]] as concurrent modes of spatial mode B
+        # q[sm[2]] as concurrent modes of spatial mode C
+        # q[sm[3]] as concurrent modes of spatial mode D.
+
+        for cmd in cmds:
+            if isinstance(cmd.op, ops.Measurement):
+                self.measured_modes.append(cmd.reg[0].ind)
+
+        for i in range(self.total_timebins):
+            for cmd in cmds:
+                self.apply_op(cmd, q, i)
+
+            if self.shift == "end":
+                # shift each spatial mode SEPARATELY by one step
+                q_aux = list(q)
+                # Here the index used to be i in changed to j, check this is fine.
+                for j in range(len(self.N)):
+                    q_aux[sm[j]] = shift_by(q_aux[sm[j]], 1)
+                q = tuple(q_aux)
+
+            elif isinstance(self.shift, int):
+                q = shift_by(q, self.shift)  # shift at end of each time bin
+
+    def apply_op(self, cmd, q, t):
+        """Apply a particular operation on register q at timestep t"""
+        params = cmd.op.p.copy()
+
+        if par_is_symbolic(params[0]):
```

Comment:
@josh146 : I believe this `if` block should be removed. I believe one should never have symbolic parameters for a sampling circuit. If you think this is the case feel free to delete the block.

Reply:
Spoke offline, but for context this is required for the TDMProgram logic. What happens internally is:

* Inside the with context, all gate arguments are replaced with sympy variables. This is used to track which parameter is used in which gate.

* After the with context exits, the TDM loop is constructed, replacing the parameters with the numeric values given by the user.